File: backend/src/controllers/paymentController.js
const { pool } = require('../config/database');
const { 
  generatePaymentId, 
  validateVPA, 
  validateCardNumber, 
  detectCardNetwork, 
  validateExpiry 
} = require('../utils/validation');

// Simulate payment processing
const processPayment = async (method, paymentId) => {
  const testMode = process.env.TEST_MODE === 'true';
  
  let delay, successRate;
  
  if (testMode) {
    // Use test mode configuration
    delay = parseInt(process.env.TEST_PROCESSING_DELAY || '1000');
    const testSuccess = process.env.TEST_PAYMENT_SUCCESS !== 'false';
    successRate = testSuccess ? 1.0 : 0.0;
  } else {
    // Use normal configuration
    const minDelay = parseInt(process.env.PROCESSING_DELAY_MIN || '5000');
    const maxDelay = parseInt(process.env.PROCESSING_DELAY_MAX || '10000');
    delay = Math.floor(Math.random() * (maxDelay - minDelay + 1)) + minDelay;
    
    successRate = method === 'upi' 
      ? parseFloat(process.env.UPI_SUCCESS_RATE || '0.90')
      : parseFloat(process.env.CARD_SUCCESS_RATE || '0.95');
  }
  
  // Simulate processing delay
  await new Promise(resolve => setTimeout(resolve, delay));
  
  // Determine success/failure
  const success = Math.random() < successRate;
  
  if (success) {
    await pool.query(
      `UPDATE payments SET status = 'success', updated_at = CURRENT_TIMESTAMP WHERE id = $1`,
      [paymentId]
    );
  } else {
    await pool.query(
      `UPDATE payments 
       SET status = 'failed', 
           error_code = 'PAYMENT_FAILED', 
           error_description = 'Payment processing failed',
           updated_at = CURRENT_TIMESTAMP 
       WHERE id = $1`,
      [paymentId]
    );
  }
};

// Create Payment
const createPayment = async (req, res) => {
  try {
    const { order_id, method, vpa, card } = req.body;
    const merchantId = req.merchant.id;

    // Validate order exists and belongs to merchant
    const orderResult = await pool.query(
      'SELECT * FROM orders WHERE id = $1 AND merchant_id = $2',
      [order_id, merchantId]
    );

    if (orderResult.rows.length === 0) {
      return res.status(400).json({
        error: {
          code: 'BAD_REQUEST_ERROR',
          description: 'Order not found or does not belong to merchant'
        }
      });
    }

    const order = orderResult.rows[0];

    // Validate method-specific fields
    let paymentData = {
      method,
      vpa: null,
      card_network: null,
      card_last4: null
    };

    if (method === 'upi') {
      if (!vpa || !validateVPA(vpa)) {
        return res.status(400).json({
          error: {
            code: 'INVALID_VPA',
            description: 'Invalid VPA format'
          }
        });
      }
      paymentData.vpa = vpa;
    } else if (method === 'card') {
      if (!card || !card.number || !card.expiry_month || !card.expiry_year || !card.cvv || !card.holder_name) {
        return res.status(400).json({
          error: {
            code: 'BAD_REQUEST_ERROR',
            description: 'Missing required card fields'
          }
        });
      }

      // Validate card number using Luhn algorithm
      if (!validateCardNumber(card.number)) {
        return res.status(400).json({
          error: {
            code: 'INVALID_CARD',
            description: 'Invalid card number'
          }
        });
      }

      // Validate expiry
      if (!validateExpiry(card.expiry_month, card.expiry_year)) {
        return res.status(400).json({
          error: {
            code: 'EXPIRED_CARD',
            description: 'Card has expired'
          }
        });
      }

      // Detect card network
      const cardNetwork = detectCardNetwork(card.number);
      const cleanedNumber = card.number.replace(/[\s-]/g, '');
      const last4 = cleanedNumber.slice(-4);

      paymentData.card_network = cardNetwork;
      paymentData.card_last4 = last4;
    } else {
      return res.status(400).json({
        error: {
          code: 'BAD_REQUEST_ERROR',
          description: 'Invalid payment method'
        }
      });
    }

    // Generate unique payment ID
    let paymentId = generatePaymentId();
    let isUnique = false;
    
    while (!isUnique) {
      const existingPayment = await pool.query('SELECT id FROM payments WHERE id = $1', [paymentId]);
      if (existingPayment.rows.length === 0) {
        isUnique = true;
      } else {
        paymentId = generatePaymentId();
      }
    }

    // Create payment with status 'processing'
    const result = await pool.query(
      `INSERT INTO payments (
        id, order_id, merchant_id, amount, currency, method, status,
        vpa, card_network, card_last4, created_at, updated_at
      )
      VALUES ($1, $2, $3, $4, $5, $6, 'processing', $7, $8, $9, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)
      RETURNING *`,
      [
        paymentId,
        order_id,
        merchantId,
        order.amount,
        order.currency,
        method,
        paymentData.vpa,
        paymentData.card_network,
        paymentData.card_last4
      ]
    );

    const payment = result.rows[0];

    // Process payment asynchronously (but wait for it to complete for this deliverable)
    await processPayment(method, paymentId);

    // Fetch updated payment
    const updatedPayment = await pool.query('SELECT * FROM payments WHERE id = $1', [paymentId]);
    const finalPayment = updatedPayment.rows[0];

    // Build response
    const response = {
      id: finalPayment.id,
      order_id: finalPayment.order_id,
      amount: finalPayment.amount,
      currency: finalPayment.currency,
      method: finalPayment.method,
      status: finalPayment.status,
      created_at: finalPayment.created_at.toISOString()
    };

    if (method === 'upi') {
      response.vpa = finalPayment.vpa;
    } else if (method === 'card') {
      response.card_network = finalPayment.card_network;
      response.card_last4 = finalPayment.card_last4;
    }

    res.status(201).json(response);
  } catch (error) {
    console.error('Error creating payment:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

// Get Payment
const getPayment = async (req, res) => {
  try {
    const { payment_id } = req.params;
    const merchantId = req.merchant.id;

    const result = await pool.query(
      'SELECT * FROM payments WHERE id = $1 AND merchant_id = $2',
      [payment_id, merchantId]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({
        error: {
          code: 'NOT_FOUND_ERROR',
          description: 'Payment not found'
        }
      });
    }

    const payment = result.rows[0];

    const response = {
      id: payment.id,
      order_id: payment.order_id,
      amount: payment.amount,
      currency: payment.currency,
      method: payment.method,
      status: payment.status,
      created_at: payment.created_at.toISOString(),
      updated_at: payment.updated_at.toISOString()
    };

    if (payment.method === 'upi') {
      response.vpa = payment.vpa;
    } else if (payment.method === 'card') {
      response.card_network = payment.card_network;
      response.card_last4 = payment.card_last4;
    }

    if (payment.error_code) {
      response.error_code = payment.error_code;
      response.error_description = payment.error_description;
    }

    res.status(200).json(response);
  } catch (error) {
    console.error('Error getting payment:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

// Create Payment (Public - for checkout page)
const createPaymentPublic = async (req, res) => {
  try {
    const { order_id, method, vpa, card } = req.body;

    // Validate order exists
    const orderResult = await pool.query('SELECT * FROM orders WHERE id = $1', [order_id]);

    if (orderResult.rows.length === 0) {
      return res.status(400).json({
        error: {
          code: 'BAD_REQUEST_ERROR',
          description: 'Order not found'
        }
      });
    }

    const order = orderResult.rows[0];

    // Validate method-specific fields
    let paymentData = {
      method,
      vpa: null,
      card_network: null,
      card_last4: null
    };

    if (method === 'upi') {
      if (!vpa || !validateVPA(vpa)) {
        return res.status(400).json({
          error: {
            code: 'INVALID_VPA',
            description: 'Invalid VPA format'
          }
        });
      }
      paymentData.vpa = vpa;
    } else if (method === 'card') {
      if (!card || !card.number || !card.expiry_month || !card.expiry_year || !card.cvv || !card.holder_name) {
        return res.status(400).json({
          error: {
            code: 'BAD_REQUEST_ERROR',
            description: 'Missing required card fields'
          }
        });
      }

      if (!validateCardNumber(card.number)) {
        return res.status(400).json({
          error: {
            code: 'INVALID_CARD',
            description: 'Invalid card number'
          }
        });
      }

      if (!validateExpiry(card.expiry_month, card.expiry_year)) {
        return res.status(400).json({
          error: {
            code: 'EXPIRED_CARD',
            description: 'Card has expired'
          }
        });
      }

      const cardNetwork = detectCardNetwork(card.number);
      const cleanedNumber = card.number.replace(/[\s-]/g, '');
      const last4 = cleanedNumber.slice(-4);

      paymentData.card_network = cardNetwork;
      paymentData.card_last4 = last4;
    } else {
      return res.status(400).json({
        error: {
          code: 'BAD_REQUEST_ERROR',
          description: 'Invalid payment method'
        }
      });
    }

    // Generate unique payment ID
    let paymentId = generatePaymentId();
    let isUnique = false;
    
    while (!isUnique) {
      const existingPayment = await pool.query('SELECT id FROM payments WHERE id = $1', [paymentId]);
      if (existingPayment.rows.length === 0) {
        isUnique = true;
      } else {
        paymentId = generatePaymentId();
      }
    }

    // Create payment
    await pool.query(
      `INSERT INTO payments (
        id, order_id, merchant_id, amount, currency, method, status,
        vpa, card_network, card_last4, created_at, updated_at
      )
      VALUES ($1, $2, $3, $4, $5, $6, 'processing', $7, $8, $9, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)`,
      [
        paymentId,
        order_id,
        order.merchant_id,
        order.amount,
        order.currency,
        method,
        paymentData.vpa,
        paymentData.card_network,
        paymentData.card_last4
      ]
    );

    // Process payment asynchronously (don't wait)
    processPayment(method, paymentId).catch(err => console.error('Payment processing error:', err));

    // Return immediate response
    const response = {
      id: paymentId,
      order_id: order_id,
      amount: order.amount,
      currency: order.currency,
      method: method,
      status: 'processing',
      created_at: new Date().toISOString()
    };

    if (method === 'upi') {
      response.vpa = paymentData.vpa;
    } else if (method === 'card') {
      response.card_network = paymentData.card_network;
      response.card_last4 = paymentData.card_last4;
    }

    res.status(201).json(response);
  } catch (error) {
    console.error('Error creating payment:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

// Get Payment (Public - for checkout page)
const getPaymentPublic = async (req, res) => {
  try {
    const { payment_id } = req.params;

    const result = await pool.query('SELECT * FROM payments WHERE id = $1', [payment_id]);

    if (result.rows.length === 0) {
      return res.status(404).json({
        error: {
          code: 'NOT_FOUND_ERROR',
          description: 'Payment not found'
        }
      });
    }

    const payment = result.rows[0];

    const response = {
      id: payment.id,
      order_id: payment.order_id,
      amount: payment.amount,
      currency: payment.currency,
      method: payment.method,
      status: payment.status,
      created_at: payment.created_at.toISOString(),
      updated_at: payment.updated_at.toISOString()
    };

    if (payment.method === 'upi') {
      response.vpa = payment.vpa;
    } else if (payment.method === 'card') {
      response.card_network = payment.card_network;
      response.card_last4 = payment.card_last4;
    }

    if (payment.error_code) {
      response.error_code = payment.error_code;
      response.error_description = payment.error_description;
    }

    res.status(200).json(response);
  } catch (error) {
    console.error('Error getting payment:', error);
    res.status(500).json({
      error: {
        code: 'INTERNAL_ERROR',
        description: 'Internal server error'
      }
    });
  }
};

module.exports = { 
  createPayment, 
  getPayment, 
  createPaymentPublic, 
  getPaymentPublic 
};

File: backend/src/main/java/com/paymentgateway/controller/PaymentController.java
package com.paymentgateway.controller;

import com.paymentgateway.dto.ErrorResponse;
import com.paymentgateway.dto.PaymentRequest;
import com.paymentgateway.dto.PaymentResponse;
import com.paymentgateway.entity.Merchant;
import com.paymentgateway.entity.Order;
import com.paymentgateway.entity.Payment;
import com.paymentgateway.repository.OrderRepository;
import com.paymentgateway.repository.PaymentRepository;
import com.paymentgateway.repository.MerchantRepository;
import com.paymentgateway.service.PaymentProcessingService;
import com.paymentgateway.util.ValidationUtil;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Optional;

@RestController
@RequestMapping("/api/v1")
public class PaymentController {
    
    @Autowired
    private PaymentRepository paymentRepository;
    
    @Autowired
    private OrderRepository orderRepository;
    
    @Autowired
    private MerchantRepository merchantRepository;
    
    @Autowired
    private PaymentProcessingService paymentProcessingService;
    
    /**
     * POST /api/v1/payments
     * Create a payment (authenticated - merchant)
     */
    @PostMapping("/payments")
    public ResponseEntity<?> createPayment(@RequestBody PaymentRequest request, HttpServletRequest httpRequest) {
        try {
            Merchant merchant = (Merchant) httpRequest.getAttribute("merchant");
            if (merchant == null) {
                return ResponseEntity.status(401).body(
                    new ErrorResponse("AUTHENTICATION_ERROR", "Invalid API credentials")
                );
            }
            
            return createPaymentInternal(request, merchant);
        } catch (Exception e) {
            System.err.println("Error creating payment: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.status(500).body(
                new ErrorResponse("INTERNAL_ERROR", "Internal server error")
            );
        }
    }
    
    /**
     * POST /api/v1/payments/public
     * Create a payment (public - checkout page)
     */
    @PostMapping("/payments/public")
    public ResponseEntity<?> createPaymentPublic(@RequestBody PaymentRequest request) {
        try {
            // Get the order to find the merchant
            Optional<Order> orderOpt = orderRepository.findById(request.order_id);
            if (orderOpt.isEmpty()) {
                return ResponseEntity.badRequest().body(
                    new ErrorResponse("BAD_REQUEST_ERROR", "Order not found")
                );
            }
            
            Merchant merchant = orderOpt.get().getMerchant();
            return createPaymentInternal(request, merchant);
        } catch (Exception e) {
            System.err.println("Error creating payment: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.status(500).body(
                new ErrorResponse("INTERNAL_ERROR", "Internal server error")
            );
        }
    }
    
    /**
     * Internal method to handle payment creation logic
     */
    private ResponseEntity<?> createPaymentInternal(PaymentRequest request, Merchant merchant) {
        // Validate order exists and belongs to merchant
        Optional<Order> orderOpt = orderRepository.findById(request.order_id);
        if (orderOpt.isEmpty()) {
            return ResponseEntity.badRequest().body(
                new ErrorResponse("BAD_REQUEST_ERROR", "Order not found or does not belong to merchant")
            );
        }
        
        Order order = orderOpt.get();
        
        // Validate payment method and method-specific fields
        if (request.method == null || !request.method.matches("^(upi|card)$")) {
            return ResponseEntity.badRequest().body(
                new ErrorResponse("BAD_REQUEST_ERROR", "Invalid payment method")
            );
        }
        
        if (request.method.equals("upi")) {
            if (request.vpa == null || !ValidationUtil.validateVPA(request.vpa)) {
                return ResponseEntity.badRequest().body(
                    new ErrorResponse("INVALID_VPA", "Invalid VPA format")
                );
            }
        } else if (request.method.equals("card")) {
            if (request.card == null || request.card.number == null || 
                request.card.expiry_month == null || request.card.expiry_year == null || 
                request.card.cvv == null || request.card.holder_name == null) {
                return ResponseEntity.badRequest().body(
                    new ErrorResponse("BAD_REQUEST_ERROR", "Missing required card fields")
                );
            }
            
            // Validate card number
            if (!ValidationUtil.validateCardNumber(request.card.number)) {
                return ResponseEntity.badRequest().body(
                    new ErrorResponse("INVALID_CARD", "Invalid card number")
                );
            }
            
            // Validate expiry
            if (!ValidationUtil.validateExpiry(request.card.expiry_month, request.card.expiry_year)) {
                return ResponseEntity.badRequest().body(
                    new ErrorResponse("EXPIRED_CARD", "Card has expired")
                );
            }
        }
        
        // Generate unique payment ID
        String paymentId;
        boolean isUnique = false;
        do {
            paymentId = ValidationUtil.generatePaymentId();
            isUnique = !paymentRepository.existsById(paymentId);
        } while (!isUnique);
        
        // Create payment entity
        Payment payment = new Payment(paymentId, order, merchant, order.getAmount(), request.method);
        
        if (request.method.equals("upi")) {
            payment.setVpa(request.vpa);
        } else if (request.method.equals("card")) {
            String cardNetwork = ValidationUtil.detectCardNetwork(request.card.number);
            String cleanedNumber = request.card.number.replaceAll("[\\s-]", "");
            String last4 = cleanedNumber.substring(cleanedNumber.length() - 4);
            
            payment.setCardNetwork(cardNetwork);
            payment.setCardLast4(last4);
        }
        
        // Save payment with 'processing' status
        paymentRepository.save(payment);
        
        // Start async payment processing
        paymentProcessingService.processPaymentAsync(paymentId, request.method);
        
        // Build response
        PaymentResponse response = new PaymentResponse();
        response.id = payment.getId();
        response.order_id = payment.getOrder().getId();
        response.amount = payment.getAmount();
        response.currency = payment.getCurrency();
        response.method = payment.getMethod();
        response.status = payment.getStatus();
        response.created_at = payment.getCreatedAt().format(DateTimeFormatter.ISO_DATE_TIME);
        
        if (request.method.equals("upi")) {
            response.vpa = payment.getVpa();
        } else if (request.method.equals("card")) {
            response.card_network = payment.getCardNetwork();
            response.card_last4 = payment.getCardLast4();
        }
        
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }
    
    /**
     * GET /api/v1/payments/{id}
     * Get payment details (authenticated - merchant)
     */
    @GetMapping("/payments/{id}")
    public ResponseEntity<?> getPayment(@PathVariable String id, HttpServletRequest httpRequest) {
        try {
            Merchant merchant = (Merchant) httpRequest.getAttribute("merchant");
            if (merchant == null) {
                return ResponseEntity.status(401).body(
                    new ErrorResponse("AUTHENTICATION_ERROR", "Invalid API credentials")
                );
            }
            
            Optional<Payment> payment = paymentRepository.findByIdAndMerchant(id, merchant);
            if (payment.isEmpty()) {
                return ResponseEntity.status(404).body(
                    new ErrorResponse("NOT_FOUND_ERROR", "Payment not found")
                );
            }
            
            return buildPaymentResponse(payment.get());
        } catch (Exception e) {
            System.err.println("Error getting payment: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.status(500).body(
                new ErrorResponse("INTERNAL_ERROR", "Internal server error")
            );
        }
    }
    
    /**
     * GET /api/v1/payments/{id}/public
     * Get payment details (public - checkout page)
     */
    @GetMapping("/payments/{id}/public")
    public ResponseEntity<?> getPaymentPublic(@PathVariable String id) {
        try {
            Optional<Payment> payment = paymentRepository.findById(id);
            if (payment.isEmpty()) {
                return ResponseEntity.status(404).body(
                    new ErrorResponse("NOT_FOUND_ERROR", "Payment not found")
                );
            }
            
            return buildPaymentResponse(payment.get());
        } catch (Exception e) {
            System.err.println("Error getting payment: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.status(500).body(
                new ErrorResponse("INTERNAL_ERROR", "Internal server error")
            );
        }
    }
    
    /**
     * Helper method to build payment response
     */
    private ResponseEntity<?> buildPaymentResponse(Payment payment) {
        PaymentResponse response = new PaymentResponse();
        response.id = payment.getId();
        response.order_id = payment.getOrder().getId();
        response.amount = payment.getAmount();
        response.currency = payment.getCurrency();
        response.method = payment.getMethod();
        response.status = payment.getStatus();
        response.created_at = payment.getCreatedAt().format(DateTimeFormatter.ISO_DATE_TIME);
        response.updated_at = payment.getUpdatedAt().format(DateTimeFormatter.ISO_DATE_TIME);
        
        if (payment.getMethod().equals("upi")) {
            response.vpa = payment.getVpa();
        } else if (payment.getMethod().equals("card")) {
            response.card_network = payment.getCardNetwork();
            response.card_last4 = payment.getCardLast4();
        }
        
        if (payment.getErrorCode() != null) {
            response.error_code = payment.getErrorCode();
            response.error_description = payment.getErrorDescription();
        }
        
        return ResponseEntity.ok(response);
    }
}
